packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.story.tsx

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright 2022 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import { MemoryRouter } from 'react-router';
+
+import { ContextProvider, Context as TeleportContext } from 'teleport';
+import cfg from 'teleport/config';
+import { ResourceKind } from 'teleport/Discover/Shared';
+import {
+  clearCachedJoinTokenResult,
+  JoinTokenProvider,
+} from 'teleport/Discover/Shared/JoinTokenContext';
+import { PingTeleportProvider } from 'teleport/Discover/Shared/PingTeleportContext';
+import { userContext } from 'teleport/mocks/contexts';
+
+import HelmChart from './HelmChart';
+
+export default {
+  title: 'Teleport/Discover/Kube/HelmChart',
+  decorators: [
+    Story => {
+      // Reset request handlers added in individual stories.
+      window.msw.worker.resetHandlers();
+      clearCachedJoinTokenResult();
+      return <Story />;
+    },
+  ],
+};
+
+export const Init = () => {
+  return (
+    <Provider>
+      <HelmChart />
+    </Provider>
+  );
+};
+
+export const Polling = () => {
+  const { worker, rest } = window.msw;
+  // Use default fetch token handler defined in mocks/handlers
+
+  worker.use(
+    rest.get(cfg.api.kubernetesPath, (req, res, ctx) => {
+      return res(ctx.delay('infinite'));
+    })
+  );
+  return (
+    <Provider>
+      <HelmChart runJoinTokenPromise={true} />
+    </Provider>
+  );
+};
+
+export const PollingSuccess = () => {
+  const { worker, rest } = window.msw;
+  // Use default fetch token handler defined in mocks/handlers
+
+  worker.use(
+    rest.get(cfg.api.kubernetesPath, (req, res, ctx) => {
+      return res(ctx.json({ items: [{}] }));
+    })
+  );
```

Comment:
Hm...any reasons not to use a high-level hook (useHelpChart.ts) here? The purpose of these hooks is to provide separation between rendering/behavior so it's easy to write stories for different states. You would also want to snapshot some of the stores and it's harder to do without access to the state. Yeah, I would keep treating stories as first-class citizens as that was **the best source of documentation.** With this we not only made it harder to storybook/shopshot-test but also introduced mocks into the environment which is not suited for that (like jest).

to summarize:  treat your storybook as documentation and do the actual testing in `jest`.

Reply:
hi @alex-kovoy!!! are you ready to come back yet? :)

You're right in that we need to write more tests, i've just been lax because the discover feature goes through so much  changes and refactoring and we also wanted it out fast. Probably a bad reason to not write tests.

This specific screen (`download/helm`) uses the [suspense](https://17.reactjs.org/docs/concurrent-mode-suspense.html) technique, which means I have to also use [error boundary](https://17.reactjs.org/docs/concurrent-mode-suspense.html#handling-errors) to gracefully handle network errors and allow for retries in the middle of the flow. 

The suspense technique doesn't use the `useAttempt` method and is a very different way to compose components, but suspense is equivalent to `attempt.status === 'loading'` and error boundary is equivalent to `attempt.status === 'failed'`. In order to see what these state looks like, i had to use `msw` and setup contexts for it. And since I was using this setup anyways, I decided to just use them for all states instead of spending extra time making dummy components. I am unsure if there is a better way to do it.

The `msw` lib allowed me to create all states possible and then the stories were used for design review.

This is the only screen that uses this technique... and I'm still on the fence about it (it makes it harder to mock and write stories), but since it was already done we decided to just go with it with the possibility of refactoring it out later or just scope usage of suspense only to this screen? Not sure b/c some devs were leery on using an experimental feature.

Other screens (that I personally create) uses the `smart/dummy` approach as you suggested. But with that said... I don't think this approach is being enforced as there are devs who does not like this approach (and I don't review all PRs). There was a whole thread of discussion on it with the result that looked like half who likes this approach (me included) and half who doesn't. And I honestly don't know what the result of that discussion was... For this particular feature, I decided that getting it out was more important.

Reply:
I am honestly unsure what our process is anymore. Are we enforcing writing stories? Are we enforcing smart/dummy approach? Right now it feels like anything goes. Maybe that's ok, as long as you can write stories for it? @hatched @ryanclark @ravicious @gzdunek @rudream @ibeckermayer @JanKaczmarkiewicz @michellescripts

Reply:
I'll try to mock out the important bits of the `download/helm` so we can snapshot it (i'm also unsure of the status of using snapshots), so then `msw` is just used to see what failed/loading state looks like and these states aren't important to snapshot.

Reply:
@kimlisa you summarize it very well why to avoid using `Suspense`. The short answer is that it's harder to test/storytell. If you want to treat your tests and stories as first-class citizens it's better to avoid Suspense. Also this App is not a static website where the basics of Suspense would suffice (hard to imagine vscode using suspense to show a loading indicator)

> The msw lib allowed me to create all states possible and then the stories were used for design review.

 Reminded me of [this commercial ](https://www.youtube.com/watch?v=g5nMbnxeQjA&ab_channel=NJMInsuranceGroup)

> There devs who does not like this approach (and I don't review all PRs). 

Devs can write tests against smart `containers` and also create stories and snapshots (rendering) against dummy `views` if needed. This separation just makes it easier (less code) to do so. Not sure what would be the reason to couple `state` and `view` together while writing tests/stories shows an opposite (basic CS concept)

Reply:
@kimlisa  
> I am honestly unsure what our process is anymore. Are we enforcing writing stories? Are we enforcing smart/dummy approach? Right now it feels like anything goes. Maybe that's ok, as long as you can write stories for it?

I would enforce writing tests + a few stories (where it makes sense) to document behavior/view (helps a lot when regression/design changes). Just the fact of treating tests and stories as _first-class citizens_ would automatically drive your decision making towards more friendly composition/architecture/less code.

@gzdunek  `term` is still missing many tests!! :) and no stories.

Reply:
> I'll try to mock out the important bits of the download/helm so we can snapshot it (i'm also unsure of the status of using snapshots), so then msw is just used to see what failed/loading state looks like and these states aren't important to snapshot.

@hatched i've decided to do this later, i'm planning to come back to refactor more stuff anyway

Reply:
> Are we enforcing writing stories? Are we enforcing smart/dummy approach? Right now it feels like anything goes. Maybe that's ok, as long as you can write stories for it?

What I was advocating for in discussions under gravitational/teleport#15813 pretty much aligns with what Alexey said here.